library/bignum.c

```diff
@@ -499,26 +499,35 @@ int mbedtls_mpi_read_string( mbedtls_mpi *X, int radix, const char *s )
 }
 
 /*
- * Helper to write the digits high-order first
+ * Helper to write the digits high-order first.
  */
-static int mpi_write_hlp( mbedtls_mpi *X, int radix, char **p )
+static int mpi_write_hlp( mbedtls_mpi *X, int radix, char **p, const size_t buflen )
 {
     int ret;
     mbedtls_mpi_uint r;
+    size_t length = 0;
+    char *p_end = *p + buflen - 1;
 
     if( radix < 2 || radix > 16 )
         return( MBEDTLS_ERR_MPI_BAD_INPUT_DATA );
 
-    MBEDTLS_MPI_CHK( mbedtls_mpi_mod_int( &r, X, radix ) );
-    MBEDTLS_MPI_CHK( mbedtls_mpi_div_int( X, NULL, X, radix ) );
+    do
+    {
+        MBEDTLS_MPI_CHK( mbedtls_mpi_mod_int( &r, X, radix ) );
+        MBEDTLS_MPI_CHK( mbedtls_mpi_div_int( X, NULL, X, radix ) );
+        /*
+         * Write the residue in the current position, as an  ASCII character.
+         */
+        if( r < 10 )
+            *p_end-- = (char)( r + 0x30 );
+        else
+            *p_end-- = (char)( r + 0x37 );
```

Comment:
This is pre-existing, but what is this line supposed to do?

Reply:
As I wrote in the comment( It took me a while to understand as well):
"Write the residue in the current position, as an  ASCII character."
Digits start from `0x30`, and `a` start from 0x41. So, 0x37 + A is 0x41

Reply:
No need to repeat a comment here. Anyway, what you wrote afterwards clarifies things, thanks.